pkg/activitypub/service/inbox/inbox.go

```diff
@@ -0,0 +1,196 @@
+	store "github.com/trustbloc/orb/pkg/activitypub/store/spi"
+	"github.com/trustbloc/orb/pkg/activitypub/vocab"
+)
+
+var logger = log.New("activitypub_service")
+
+type pubSub interface {
+	Subscribe(ctx context.Context, topic string) (<-chan *message.Message, error)
+	Publish(topic string, messages ...*message.Message) error
+	Close() error
+}
+
+// Config holds configuration parameters for the Inbox.
+type Config struct {
+	ServiceName   string
+	ListenAddress string
+	Topic         string
+}
+
+// Inbox implements the ActivityPub inbox.
+type Inbox struct {
+	*Config
+	*lifecycle.Lifecycle
+
+	router          *message.Router
+	httpSubscriber  *wmhttp.Subscriber
+	msgChannel      <-chan *message.Message
+	activityHandler service.ActivityHandler
+	activityStore   store.Store
+	jsonUnmarshal   func(data []byte, v interface{}) error
+}
+
+// New returns a new ActivityPub inbox.
+func New(cfg *Config, s store.Store, pubSub pubSub, activityHandler service.ActivityHandler) (*Inbox, error) {
+	h := &Inbox{
+		Config:          cfg,
+		activityHandler: activityHandler,
+		activityStore:   s,
+		jsonUnmarshal:   json.Unmarshal,
+	}
+
+	h.Lifecycle = lifecycle.New(cfg.ServiceName, h.start, h.stop)
+
+	msgChan, err := pubSub.Subscribe(context.Background(), cfg.Topic)
+	if err != nil {
+		return nil, errors.WithMessagef(err, "unable to subscribe to topic [%s]", cfg.Topic)
+	}
+
+	httpSubscriber, err := wmhttp.NewSubscriber(cfg.ListenAddress, wmhttp.SubscriberConfig{}, wmlogger.New())
+	if err != nil {
+		return nil, errors.WithMessagef(err, "unable to create HTTP subscriber")
+	}
+
+	router, err := message.NewRouter(message.RouterConfig{}, wmlogger.New())
+	if err != nil {
+		return nil, errors.WithMessagef(err, "unable to create router")
+	}
+
+	router.AddMiddleware(middleware.Recoverer, middleware.CorrelationID)
+
+	router.AddPlugin(plugin.SignalsHandler)
+
+	router.AddHandler(
+		"inbox-"+cfg.ServiceName, cfg.ServiceName,
+		httpSubscriber, cfg.Topic, pubSub,
+		func(msg *message.Message) ([]*message.Message, error) {
+			// Simply forward the message.
+			return message.Messages{msg}, nil
+		},
+	)
+
+	h.router = router
+	h.httpSubscriber = httpSubscriber
+	h.msgChannel = msgChan
+
+	return h, nil
+}
+
+func (h *Inbox) start() {
+	// Start the router
+	go h.route()
+
+	// Start the message listener
+	go h.listen()
+
+	// HTTP server needs to be started after router is ready.
+	<-h.router.Running()
+
+	// Start the HTTP server
+	go h.serveHTTP()
+}
+
+func (h *Inbox) stop() {
+	if err := h.httpSubscriber.Close(); err != nil {
+		logger.Warnf("Error closing inbox: %s", err)
+	}
+
+	if err := h.router.Close(); err != nil {
+		logger.Warnf("[%s] Error closing router: %s", h.ServiceName, err)
+	} else {
+		logger.Debugf("[%s] Closed router", h.ServiceName)
+	}
+}
+
+func (h *Inbox) route() {
+	logger.Debugf("[%s] Starting router", h.ServiceName)
+
+	if err := h.router.Run(context.Background()); err != nil {
```

Comment:
What does this context get used for?

Reply:
Looks like it's used by the router to subscribe to the 'subscriber' that's passed in. I guess persistent message queues require a context.